### DISCUSSION TITLE
How to split specified modules and their dependencies into one bundle

### DISCUSSION THREAD

**Discussion by curly210102** [webpack/webpack#13917] [Q&A]
The main entry `index.js` import module `a` and `b` under the "components" directory, and module `a` import module `c`, `index.js` and `c` are not related directly. I want to split the imported modules under "components", bundle them with the modules they depend on. Below is the webpack config.

``` js
const path = require("path");
const BundleAnalyzerPlugin = require("webpack-bundle-analyzer")
  .BundleAnalyzerPlugin;

module.exports = {
  mode: "production",
  entry: "./src/index.js",
  output: {
    path: path.resolve(__dirname, "dist")
  },
  optimization: {
    runtimeChunk: "single",
    splitChunks: {
      cacheGroups: {
        vendor: {
          test: /[\\/]components[\\/]/,
          name: "vendors",
          chunks: "all",
          priority: 10,
          enforce: true,
          reuseExistingChunk: true
        }
      }
    }
  },
  plugins: [
    new BundleAnalyzerPlugin({
      openAnalyzer: false
    })
  ]
};

```

You can reproduce it in https://codesandbox.io/s/silly-violet-k6gew?file=/webpack.config.js

![image](https://user-images.githubusercontent.com/76864176/128015652-f07a438f-1c69-4e61-b7e8-ceb10f3b5c50.png)

As shown in the analysis, module `c` is splitted into the main bundle. I expect the custom group have higher priority to inculde depend on modules, how to achieve it?

Thanks for help ❤️

**Comment by alexander-akait** [comment]
You can use `import()` for this case, do you use react or vue or something else?

**Reply by curly210102** [reply]
Hi, thanks for reply, and yes I was tring to optimize a site which is built with react and a UI Library.
I was tring to split the UI components and site code, because the dependencies of UI components  occupies 2/3 of node_modules.

After asking this question, I read the document again and have new ideas:
- "Multi entry points", maybe I can maintain a module to import and export the UI components which are in use.
- "import()", It's smart, but one bundle per component, It'will increase lots of request and can't prevent duplication
- "cacheGroup": I guess the design intent of cacheGroup is to extract specific modules themselves, It won't traverse the dependency graph, the feature is provided by entry points(static) and import(dynamic)

As a result, If using little UI components, `import()` will be good choice, but if there are a lot, choose "multi entry points".

That is all my thought.

**Reply by samagragupta** [reply]
So what would be the best solution

**Reply by samagragupta** [reply]
@curly210102  Also how will not include c.js in a.js?
